python/pylibraft/pylibraft/neighbors/cagra/cagra.pyx

```diff
@@ -0,0 +1,1148 @@
+
+from pylibraft.common.cpp.mdspan cimport (
+    device_matrix_view,
+    device_vector_view,
+    make_device_vector_view,
+    row_major,
+)
+from pylibraft.common.mdspan cimport (
+    get_dmv_float_uint32,
+    get_dmv_int8_uint32,
+    get_dmv_int64_uint32,
+    get_dmv_uint8_uint32,
+    get_dmv_uint32_uint32,
+    get_hmv_float_uint32,
+    get_hmv_int8_uint32,
+    get_hmv_int64_uint32,
+    get_hmv_uint8_uint32,
+    get_hmv_uint32_uint32,
+    make_optional_view_int64,
+)
+from pylibraft.neighbors.common cimport _get_metric_string
+from pylibraft.neighbors.ivf_pq cimport ivf_pq
+from pylibraft.neighbors.ivf_pq.cpp.c_ivf_pq cimport (
+    index_params as IVFPQ_IP,
+    search_params as IVFPQ_SP,
+)
+
+
+cdef class IndexParams:
+    cdef c_cagra.index_params params
+
+    def __init__(self, *,
+                 metric="sqeuclidean",
+                 intermediate_graph_degree=128,
+                 graph_degree=64,
+                 add_data_on_build=True):
+        """"
+        Parameters to build index for IVF-PQ nearest neighbor search
+
+        Parameters
+        ----------
+        metric : string denoting the metric type, default="sqeuclidean"
+            Valid values for metric: ["sqeuclidean", "inner_product",
+            "euclidean"], where
+            - sqeuclidean is the euclidean distance without the square root
+              operation, i.e.: distance(a,b) = \\sum_i (a_i - b_i)^2,
+            - euclidean is the euclidean distance
+            - inner product distance is defined as
+              distance(a, b) = \\sum_i a_i * b_i.
+        intermediate_graph_degree : int, default = 128
+
+        graph_degree : int, default = 64
+
+        add_data_on_build : bool, default = True
+            After training the coarse and fine quantizers, we will populate
+            the index with the dataset if add_data_on_build == True, otherwise
+            the index is left empty, and the extend method can be used
+            to add new vectors to the index.
+        """
+        self.params.metric = _get_metric(metric)
+        self.params.metric_arg = 0
+        self.params.intermediate_graph_degree = intermediate_graph_degree
+        self.params.graph_degree = graph_degree
+        self.params.add_data_on_build = add_data_on_build
+
+    @property
+    def metric(self):
+        return self.params.metric
+
+    @property
+    def intermediate_graph_degree(self):
+        return self.params.intermediate_graph_degree
+
+    @property
+    def graph_degree(self):
+        return self.params.graph_degree
+
+    @property
+    def add_data_on_build(self):
+        return self.params.add_data_on_build
+
+
+cdef class Index:
+    cdef readonly bool trained
+    cdef str active_index_type
+
+    def __cinit__(self):
+        self.trained = False
+        self.active_index_type = None
+
+
+cdef class IndexFloat(Index):
+    cdef c_cagra.index[float, uint32_t] * index
+
+    def __cinit__(self, handle=None):
```

Comment:
Not necessarily in the current release, but as a follow up work:

I would recommend that we have a constructor that takes two arrays (dataset, and graph) and just wraps them as an index object, like here: https://github.com/rapidsai/raft/blob/4574e9aaf68e8cb547d417f60359502d6bcbd94b/cpp/include/raft/neighbors/cagra.cuh#L268

This way we can experiment better with different index configurations. Here is an example workflow

```python
dataset = load_dataset(filename) # loads dataset on host
# it is efficient to build graph using host input array
knn_graph = build_knn_graph(dataset, ...) 

# One could use any method to bulid knn_graph.
# It might be needed to sort the neighbors, if it is not yet done.
# sort_knn_graph(knn_graph) # this is unnecessary if build_knn_graph was used

opt_graph = np.empty((n_rows, graph_degree)
opt_graph = optimize(knn_graph, opt_graph)

# if dataset is correctly aligned, then it makes sense to already move to GPU
# the index will only store a reference to it, and we can keep multiple indices in memory (assuming graphs fit)
dataset = cp.asarray(dataset) 
index = Index(dataset, opt_graph)
distances, neighbors = search(index, queries, k)

# Try another variation, using the same knn_graph, but different optimization options
opt_graph2 = np.empty(n_rows, other_degree)
opt_graph2 = optimize(knn_graph, opt_graph2)

index2 = Index(dataset, opt_graph2)

# search with index2, compare qps / recall to index
```

Reply:
Sounds good, I think we can do a follow up PR to add that and testing for it, so for this one will remove the functions so that we keep the released API consistent and simple to use